### ISSUE TITLE
[API STABILITY] Are we sure about "ofDoubles" trick in the metrics builders vs duplicate API?

### ISSUE THREAD

**Issue by bogdandrutu** [open-telemetry/opentelemetry-java#3988]
The benefit of not having an extra builder method on [Meter](https://github.com/open-telemetry/opentelemetry-java/blob/main/api/all/src/main/java/io/opentelemetry/api/metrics/Meter.java#L49) VS the hidden trick [ofDoubles](https://github.com/open-telemetry/opentelemetry-java/blob/main/api/all/src/main/java/io/opentelemetry/api/metrics/LongCounterBuilder.java#L30) does not make sense.

Should we consider to have a simple API "doubleCounterBuilder" and "longCounterBuilder".

There are other inconsistency since for some types we return the "Long" as the default builder (for Counter), vs others where we return "Double" (for Gauge).

I think we should block the release of 1.10 and solve at least the inconsistent return types, and preferable remove the trick of "ofDoubles"

**Comment by bogdandrutu** [comment]
Another approach would be to really have a "CounterBuilder" and "GaugeBuilder" that will have only 2 methods "ofLongs" and "ofDoubles" if too many methods on the Meter confuse users.

**Comment by anuraaga** [comment]
IIRC, the idea was that it's quite uncommon to use double counters or long gauges, so the API provides the common types as simply as possible. I think @jsuereth had discussed this with the metrics SIG and there was agreement that these default types make sense. I personally like the smaller entrypoint for common metrics usage. @jsuereth Do you have any context to add?

**Comment by jsuereth** [comment]
@anuraaga nailed the discussion and reason behind this choice.  I support the current form, which incidentally arose from a comment you made on the early prototype and subsequent conversation with Java SiG.

TL;DR on use cases:

- Counters -> count things, likely integers
- Histograms/Gauge -> measure things, likely floating point values.

If we provide a timer instrument, that's likely default to integers.

**Comment by jkwatson** [comment]
I also prefer the current API and wouldn't change it.

**Comment by bogdandrutu** [comment]
Thanks for adding this comments. Since I am not that involved, I will defer the decisions to you folks, so up to you, but seriously this API smells a bit 😄

**Comment by jsuereth** [comment]
We discussed internally, but I'll let this one simmer for discussion for one SiG.  This API was *already* the result of discussion among maintainers.  All of the points you raised were already raised during that discussion, so I doubt we'd change our minds without any additional  input / points.

**Comment by bogdandrutu** [comment]
@jsuereth what about the proposal to add an extra step for all:

```java
public interface GaugeBuilder {
  /** Sets the gauge for recording {@code long} values. */
  LongGaugeBuilder ofLongs();

  /** Sets the gauge for recording {@code double} values. */
  DoubleGaugeBuilder ofDoubles();
}
```

The main reason for doing this is because we don't know if in the future we may offer possibility to record "Strings" or "Enum" for Gauges at least (the list can grow and having all of these on the DoubleGauge does not make too much sense). I've seen backends like Stackdriver (and others) which offers the possibility to record other non numerical types at least for Gauges, and if we are doing this for Gauge I think we should do for all.

**Comment by jsuereth** [comment]
@bogdandrutu The use case of non-default types being recorded already fits in the current proposal.   We default to `double` for Gauge, and if we  do support Strings or Enums they're unlikely to be the common case.  Even for backends which support them (like Google Cloud Monitoring, formerly Stackdriver), we don't see it as common, in fact OpenTelemetry/OpenCensus must [emulate them anyway](https://cloud.google.com/monitoring/api/v3/kinds-and-types#emulate-string-vals).

I'm glad you added a counter-proposal for evaluation.  This was one of the options we had previously discussed.  IIUC your concerns are:

1. Picking a default for the builder and only requiring `ofOtherType` is hard to discover.
2. The fact we have different defaults for instrument types between instruments is "a smell".

Regarding (1) I think that's something we can discuss/identify and figure out how much of an issue it is in practice.  Specifically, if we think users frequently want/need `ofOtherType` then our defaults are pretty poor.   If, as we suspect, it's a 20% use case and users are unable to find these in the builder, I'd suspect we have more issues with our builder-pattern in general (thinking about SchemaURL).   Put another way, I'm not sure a world where `ofDoubles` and `ofLongs` didn't exist would be so bad.  They exist now because there is no alternative we can give users who need long precision.

Regarding (2) I think what we've done is in line with OTEL metrics vs. OpenCensus.   The idea behind multiple instruments is to streamline specific use cases vs. "I record doubles" or "I record longs" and then forcing users to derive what they need from those primitives.   

If we take inspiration from Micrometer (which we do), Gauges default to doubles,  Histograms are more flexible, but we expect them to be used in Timer scenarios, which in Micrometer appropriately uses `Duration` or `long` + `TimeUnit` (or just does the timing on your behalf).   The odd duck here is [Counter](https://javadoc.io/doc/io.micrometer/micrometer-core/latest/io/micrometer/core/instrument/Counter.html) where micrometer allows incrementing by doubles, but defaults the `increment()` method to add one.  So here, I think we're inline with common usage of counters.

Now, let's talk about the downsides to what you propose:

- Our APIs become more verbose for common usage.   We already have to deal with MeterProvider -> Meter -> Instrument when most other libraries have Meter -> Instrument.  Additionally, those libraries aren't forcing the representation choice on users.  This is one of those decisions that adds up in a LOT of places.  One reason we moved from `longCounterBuilder` to just `counterBuilder` is because users don't think "I want to record longs", they think "I want to count XYZ".
- It sets a bad precedence for use-case-optimized instruments (e.g. Timers, the BIG lacking instrument in our API).
- In practice, long vs. double decision has little practical importance to users, but we force them to decide.   E.g. when exporting to prometheus this distinction disappears.  The only case where we (at least me) care a lot is for cumulative counters and when integer-roll-over occurs.

**Comment by bogdandrutu** [comment]
> Specifically, if we think users frequently want/need ofOtherType then our defaults are pretty poor. If, as we suspect, it's a 20% use case and users are unable to find these in the builder, I'd suspect we have more issues with our builder-pattern in general (thinking about SchemaURL). Put another way, I'm not sure a world where ofDoubles and ofLongs didn't exist would be so bad. They exist now because there is no alternative we can give users who need long precision.

When are you planing to do this research and understand if our defaults are good or bad? After stable API? What is the backup plan if we made a mistake?

> If we take inspiration from Micrometer (which we do), Gauges default to doubles, Histograms are more flexible, but we expect them to be used in Timer scenarios, which in Micrometer appropriately uses Duration or long + TimeUnit (or just does the timing on your behalf).

Good, even though I think we kind of take more inspiration from Prometheus, at least I believe (I may be wrong) Prometheus is way more used overall than micrometer, there Counters are doubles by default.

> Our APIs become more verbose for common usage. We already have to deal with MeterProvider -> Meter -> Instrument when most other libraries have Meter -> Instrument. Additionally, those libraries aren't forcing the representation choice on users. This is one of those decisions that adds up in a LOT of places. One reason we moved from longCounterBuilder to just counterBuilder is because users don't think "I want to record longs", they think "I want to count XYZ".

I still believe that the initial API as it was initially is good enough without too many redirects. I heard that the argument was that it is confusing for users, but looking at things like micrometer I see 10+ types (are these not confusing?). I am confused about the motivation of changing the initial API.

**Comment by anuraaga** [comment]
We discussed some more today and have decided to stick to using ofLongs / ofDoubles.

- The current APIs have been out for several months, and used in several projects, including non-maintainers, for example in jfr-streaming, and we haven't heard any feedback that the long/double defaults are causing UX problems (UpDownCounter on the other hand... :P).

- It seems unlikely we could provide a highly consistent metrics API experience across languages. Java's builder pattern already differentiates it significantly, and enables the type-switching experience we found, whereas it's unlikely Go could implement a similar technique. Thanks to that, though, it seems that Java's metrics API is closer to JS and Python, which have no concept of long / double in their APIs from what I can tell.
  - Though we do have the difference where counters default to longs, while in those languages I think all default to floats. @jsuereth The one change that we could consider again is to make all of our instruments default to double, including counters. This would be consistent with opentelemetry-js, -python, micrometer, prometheus which seems like a lot of precedent. It's a scary last-second change though

- If we do find problems later, we could add `doubleCounter` / `longCounter` and deprecate the current ones. It's not ideal, but with all the discussion including what's in this issue, we would prefer to try the current style.

**Comment by bogdandrutu** [comment]
> The one change that we could consider again is to make all of our instruments default to double, including counters. This would be consistent with opentelemetry-js, -python, micrometer, prometheus which seems like a lot of precedent. It's a scary last-second change though

This will still improve in my opinion a lot the consistency of the API itself and with previous implementations. I would support this change.

**Comment by jsuereth** [comment]
My thoughts here:

- Agreed that UpDownCounter may be the most problematic abstraction n our API.
- Regarding consistency in APIs across languages, I've been doing a lot of TC trace reviews lately (C++, erlang) and have become familiar with things across languages.   Unfortunately, I think the level of consistency you're talking about just doesn't exist in any OTel SDK.   One of the principles in consistency is in the *abstractions* and terms.  A Counter should be a Counter in any language.  However, languages *should* optimise for their idioms and style.   We saw the trace spec loosen some of its interface lingo as implementations fleshed out.  The builder pattern in Java separates it, not just in Metrics but also Traces.
- I agree if we find problems we can split to `doubleCounter`/`longCounter` later.

Let's talk about a deeper issue though:

>  I think all default to floats.

OTel's decision to differentiate `integer` measurements and `floating point` measurements.  Let's look at critical Java metric instrumentation that we absolutely should to be compatible with:

- [Dropwizard metrics](https://metrics.dropwizard.io/4.2.0/manual/core.html) 
  - also allows differentiated measurements for gauges.
  - Fixes [counters to only be int64](https://metrics.dropwizard.io/4.2.0/manual/core.html#counters)
  - Histograms default to *long* or *int* measurements
- [Micrometer metrics](https://javadoc.io/doc/io.micrometer/micrometer-core/1.0.3/io/micrometer/core/instrument/Metrics.html)
  - Store all their own metrics/storage in floating point numbers.
  - API
    - Allow any `T extends Number` for Gauge
    - Counter does not specify representation but allows double or "increment" inputs.   
    - When interfacing with DropWizard Micrometer [silently forces all double increments to be integers](https://github.com/micrometer-metrics/micrometer/blob/main/micrometer-core/src/main/java/io/micrometer/core/instrument/dropwizard/DropwizardCounter.java#L35)
- Prometheus Client Library
  - Defaults everything to input doubles
  - Has an `increment()` method on Counter, similar to Micrometer that acts like integer ++.
  - Has a different input for "Timer" where you're interacting with the instrument in a completely different way.
- OpenCensus Stats
  - Requires users to pick integer vs. floating point FIRST.  There are only "two" instrument types, MeasureDouble + MeasureLong.
  - Requires user to select resulting metric only via VIEWs (common source of confusion/questions/issues we see).
  
Two important points here:

1. We *should not aim to replace either Micrometer or Dropwizard or Prometheus*.  Instead we should aim to bridge these APIs where we can, and allow OTEL signals to enhance them for correlation (e.g. adding exemplars, attaching to Resource).  As such, our backend supports both integer + floating point data storage, and our API *must* expose hooks for bridging with these APis.  We should not think of these hooks as primary use case for the API, but definitely high-priority features.
2. If users *do* choose to adopt our API for metrics (because they already use trace and want "one dependency", e.g.) then we should give them the best API we can, with the best default we know.  Given the differences in instruments (and expected input types) across every Metric API I've looked into (both Java and non-Java)



So given #1 - I don't think we should get rid of Long vs. Double in our backend storage (although it's something we've considered several times).  Additionally, it means we can't just remove `ofXYZ` from our builders for compatibility reasons (although that'd be my preference).

For #2 - I still think integer is the most natural and consistent experience for Counter. We have not heard any negative user feedback on it, and folks who have tried it are *more* hung up on Gauge vs. UpDownCounter than anything else.  The benefits of default integer counter:
- More accurate rates for long-running processes.  Double counters suffer at higher numbers with low-rates where you can get into a state where you're no longer able to represent accurate counts.  The failure scenario here is well worse than integers (where resets/rollovers are generally expected and handled query-side in storage).
- It's in line w/ DropWizard and matches most of our semantic conventions/usage right now.

TL;DR: I'm still not convinced double is a better default of Counter nor that this is a significant cause of friction for users.   I think our time would be better spent discussing UpDownCounter.   I vote against changing the default from integer => floating point for counter, especially as it seems like needless churn in an API that hasn't caused confusion.

That said, if Java Maintainers would prefer everything to default to double, that's a reasonable standpoint (even if I disagree), and still accomplishes the most important features the API needs to provide.

**Comment by jkwatson** [comment]
I would still strongly prefer keeping the current API.

**Comment by jsuereth** [comment]
Going to close this as won't fix.
